### ISSUE TITLE
Releasing v4.0.0?

### ISSUE THREAD

**Issue by kevinlewi** [dalek-cryptography/curve25519-dalek#405]
Hello! Just wondering if there is a timeline or any blockers for upgrading the pre-release (v4.0.0-pre.1 currently) to v4.0.0. The opaque-ke library (https://github.com/novifinancial/opaque-ke) is also currently on a pre-release, depending on v4.0.0-pre.1 of this library, and I am waiting on this dependency to upgrade to a full release before doing the same with opaque-ke.

Thanks!

**Comment by tarcieri** [comment]
Another issue it'd probably be good to address before a v4.0 release: https://github.com/dalek-cryptography/curve25519-dalek/pull/394/files

**Comment by pinkforest** [comment]
Fixing all the broken links and moving to docs.rs hosting #439 fixing #341

**Comment by tarcieri** [comment]
It looks like all of the issues in the list I made are fixed, however I opened one more: #442

**Comment by tarcieri** [comment]
#449 would be good to address as well

**Comment by pinkforest** [comment]
The two issues have been addressed - 

Also re: #449 I'm trying to sneak in these 2x PR's that will solve it:

- https://github.com/dalek-cryptography/curve25519-dalek/pull/463
- https://github.com/dalek-cryptography/curve25519-dalek/pull/465

#449 is otherwise sorted as the cfg(curve25519_dalek_bits = "32"|"64") was added in #454

**Comment by tarcieri** [comment]
The entire checklist is checked off. Seems like aside from e.g. documentation polish there are no remaining blockers for a final release?

**Comment by pinkforest** [comment]
I just need to fix the build.rs for wasm and cross-targets

**Comment by pinkforest** [comment]
Also the backend selection documentation is here now ready:
 - https://github.com/dalek-cryptography/curve25519-dalek/pull/467

**Comment by SergeStrashko** [comment]
Releasing?

**Comment by tarcieri** [comment]
#489 is another change that would be nice to get in before a breaking release. I went ahead and added it to the list, and tagged it `do-for-4.0`

**Comment by nox** [comment]
Seems like #489 has been handled and all that remains is #491.

**Comment by jplatte** [comment]
Is there anything the community can do to make the release happen`

**Comment by tarcieri** [comment]
#380 looks like it might be worth addressing

**Comment by rozbb** [comment]
@jplatte we have some prereleases on the 4.0 track already. If you use `4.0.0-rc.2` everything should Just Work.

**Comment by rozbb** [comment]
@jplatte x25519-dalek rc.2 is out

**Comment by tarcieri** [comment]
As far as I can tell all of the issues/PRs for proposed breaking changes are now closed: #521

I think #523 would be good to land first before a release, ideally getting another prerelease to make sure it doesn't break anything. I've dealt with miscompilation issues in the past with runtime feature autodetection + SIMD code so it'd be really nice to have some confidence that's not going to happen.

If we can land that and get another release candidate out, I think we're otherwise finally ready to release.

cc @rozbb @koute

**Comment by koute** [comment]
>  I've dealt with miscompilation issues in the past with runtime feature autodetection + SIMD code so it'd be really nice to have some confidence that's not going to happen.

FWIW FYI for the SIMD autodetection I replaced our use of `curve25519-dalek` in [substrate](https://github.com/paritytech/substrate) + [polkadot](https://github.com/paritytech/polkadot/) and synced our chain from scratch, both using the AVX2 and the AVX512 backends, and it correctly synced everything. So at the very least it shouldn't be horribly broken.

**Comment by daira** [comment]
It looks like this crate has been stuck on 4.0.0-rc.2 since March. That's unusually long for an rc period. A release would help us (zcashd) to eliminate old versions of `digest` and `rand-core` as dependencies.

**Comment by tarcieri** [comment]
#523 is the main blocker, and it's ready to merge

**Comment by rozbb** [comment]
Agreed. Sorry, end of semester and beginning of summer internship has been a lot. I think we can cut a new release this week, and maybe the other dalek crates at the same time (there's just 1 outstanding PR in ed- I think that upgrades to the new scalar API)

**Comment by tarcieri** [comment]
I reopened #414 to discuss the new feature-gating changes in #523, in hopes we can again simplify that back to something like what it had before, as I feel things have gotten problematically complex again with no clear use cases for that additional complexity

**Comment by hrxi** [comment]
It seems #414 is closed again. Does this mean a release could happen?

**Comment by AaronFeickert** [comment]
Any thoughts on #546? I opened it just a few hours after the last comment.

**Comment by jrose-signal** [comment]
I see there are still questions about ed25519-dalek, but anything holding up a release of x25519-dalek in the meantime? We use x+curve but not ed.
